### PR TITLE
New API function to retrieve members of a GroupItem

### LINKED ISSUE #586: Get all members including group items
migrated from Bugzilla [#465201](https://bugs.eclipse.org/bugs/show_bug.cgi?id=465201)
status NEW severity _enhancement_ in component _Core_ for _---_
Reported in version _unspecified_ on platform _PC_
Assigned to: Project Inbox

On 2015-04-22 10:11:04 -0400, Smart Home wrote:

> Currently getAllMembers() returns all members of group item excluding group items.
> Can it be extended (new function?) to support returning all memebers including group items?

On 2015-04-27 09:09:45 -0400, Dennis Nobel wrote:

> We can not change the existing method because of backward-compatibility, but we can introduce another method. Do you need in in the Java API or would the REST API be sufficient?

On 2015-04-28 15:29:29 -0400, Smart Home wrote:

> I need it in the Java API as rest API is not supported with my application.
> 
> Do you plan to add something like predicates in the future?

On 2015-04-28 16:14:18 -0400, Dennis Nobel wrote:

> I think this is related to the discussion in the forum: https://www.eclipse.org/forums/index.php/m/1692318/?srch=guava#msg_1692318 ?
> 
> I think the concept of predicates can be supported, but we need to decide wether we want to have Guava dependencies or we want to define our own interfaces.

On 2015-04-28 16:30:02 -0400, Smart Home wrote:

> Maybe we can start by defining the same interface Guava have for predicates and in the future replace it if Guava will be used. 
> 
> Starting using predicates now will save the need to add different functions of filtering and will be easy to replace by official predicates in the future.

On 2015-04-29 12:35:42 -0400, Dennis Nobel wrote:

> > Maybe we can start by defining the same interface Guava have for predicates and in the future replace it if Guava will be used. 
> 
> This would break the API, if we later change to Guava.
> 
> Guava is already used in the internal implementation of Eclipse SmartHome, but it is not yet part of the API. As we tend to have a clean API without dependencies to third party libs, we should better define our own interface and see how it can be combined with Guava.

### PR DESCRIPTION
**Pull request by triller-telekom**
With Java 8 we can now use streams and their filter functions, so this method can retrieve any members matching a given filter

Fixes #586

Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GroupItemTest.groovy [hunk 1]
```diff
@@ -92,6 +94,55 @@ class GroupItemTest extends OSGiTest {
...
+                    @Override
+                    public boolean test(Item t) {
+                        if(t instanceof GroupItem) {
+                            return true;
+                        }
+                        return false;
+                    }
+                };
+
+        Set<Item> members = rootGroupItem.getMembers(groupItemsInMembers)
```
Comment (triller-telekom) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GroupItemTest.groovy]: Just figured out they do, but they have a slightly different syntax. Works now.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 2]
```diff
@@ -109,22 +111,36 @@ public GroupFunction getFunction() {
...
+        Set<Item> allMembers = getMembers((Item i) -> !(i instanceof GroupItem));
         return ImmutableSet.copyOf(allMembers);
     }
 
     private void collectMembers(Set<Item> allMembers, Set<Item> members) {
         for (Item member : members) {
             if (member instanceof GroupItem) {
                 collectMembers(allMembers, ((GroupItem) member).members);
+                allMembers.add(member);
+
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: Have you added this empty line by intention?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 3]
```diff
@@ -109,22 +111,36 @@ public GroupFunction getFunction() {
      * @return all members of this and all contained {@link GroupItem}s
      */
     public Set<Item> getAllMembers() {
-        Set<Item> allMembers = new HashSet<Item>();
-        collectMembers(allMembers, members);
+        Set<Item> allMembers = getMembers((Item i) -> !(i instanceof GroupItem));
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: Now we can also use a one liner...

WRT: http://stackoverflow.com/a/31734226/7546824

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java (hunk 2 above)
Comment (triller-telekom) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: No, this unfortunately just happens to me sometimes. I create them to get a better overview while coding and moving lines around and then forget to clean them up afterwards. I have removed it now.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java (hunk 3 above)
Comment (triller-telekom) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: Indeed, I have removed the creation of the local variable there.

**Comment by triller-telekom** [comment]
Arrrg travis, c'mon you can download the bundles... try harder...

**Comment by kaikreuzer** [comment]
restarted Travis...

**Pull request merged**